project/part5/src/main.rs

```diff
@@ -0,0 +1,723 @@
+        ModeConfig::AccessPoint(AccessPointConfig::default().with_ssid("esp-radio".into()));
+    controller
+        .set_config(&ap_config)
+        .expect("Failed to set AP WiFi configuration");
+    info!("Starting WiFi in AP mode");
+    controller
+        .start_async()
+        .await
+        .expect("Failed to start WiFi");
+    debug!("WiFi AP started!");
+
+    // Wait for credentials
+    debug!("Waiting for WiFi credentials...");
+    let credentials = wifi_credentials_channel.receiver().receive().await;
+    info!("Credentials received! SSID: {}", credentials.ssid);
+
+    // Give the HTTP handler time to send the saved page before dropping AP
+    debug!("Delaying AP shutdown to allow HTTP response to complete...");
+    Timer::after(EmbassyDuration::from_secs(2)).await;
+
+    // Stop the AP
+    debug!("Stopping AP mode...");
+    controller.stop_async().await.expect("Failed to stop WiFi");
+    debug!("AP stopped");
+
+    Timer::after(EmbassyDuration::from_secs(1)).await;
+
+    // Configure and start station mode
+    debug!("Configuring station mode...");
+    let client_config = ClientConfig::default()
+        .with_ssid(credentials.ssid.as_str().into())
+        .with_password(credentials.password.as_str().into());
+
+    let sta_config = ModeConfig::Client(client_config);
+    controller
+        .set_config(&sta_config)
+        .expect("Failed to set station mode WiFi configuration");
+
+    debug!("Starting WiFi in station mode...");
+    controller
+        .start_async()
+        .await
+        .expect("Failed to start WiFi");
+    debug!("WiFi station started!");
+
+    // Connect to the network
+    debug!("Connecting to WiFi network...");
+    loop {
+        match controller.connect_async().await {
+            Ok(()) => {
+                debug!("Successfully connected to WiFi!");
+
+                // Wait for disconnect event
+                controller.wait_for_event(WifiEvent::StaDisconnected).await;
+                debug!("Disconnected from WiFi, will attempt to reconnect...");
+            }
+            Err(e) => {
+                error!("Failed to connect: {:?}", e);
+                debug!("Retrying in 5 seconds...");
+                Timer::after(EmbassyDuration::from_secs(5)).await;
+            }
+        }
+    }
+}
+
+#[embassy_executor::task]
+async fn net_task(mut runner: Runner<'static, WifiDevice<'static>>) {
+    runner.run().await
+}
+
+#[embassy_executor::task]
+async fn sta_net_task(mut runner: Runner<'static, WifiDevice<'static>>) {
+    runner.run().await
+}
+
+#[embassy_executor::task]
+async fn mqtt_task(
+    stack: Stack<'static>,
+    mut sht: ShtC3<esp_hal::i2c::master::I2c<'static, esp_hal::Async>>,
+) {
+    // Wait for WiFi connection
+    debug!("MQTT: Waiting for WiFi link to come up...");
+    stack.wait_link_up().await;
+    debug!("MQTT: WiFi link up, waiting for network configuration...");
+
+    // Wait for DHCP to assign an IP address
+    loop {
```

Comment:
I don't think we need to do this in every task. Plus the link/connection can go up or down at anytime, so if we're really going to add this check it should be part of the tasks control flow.